### PR TITLE
Add Dynamic Walker. Improved Ragdoll Stability/Performance

### PR DESCRIPTION
**Pull request by Hunter-Unity**
### Proposed change(s)

Removed the Walker environment and added WalkerDynamic and WalkerStatic environments. The ragdoll has been rebuilt to improve stability & performance.

### Types of change(s)

- [ x] Bug fix
- [ x] New feature
- [ x] Code refactor
- [ ] Breaking change
- [ x] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ x] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+    public class DirectionIndicator : MonoBehaviour
+    {
+        public Transform transformToFollow; //ex: hips or body
+        public Transform targetToLookAt; //target in the scene the indicator will point to
+        public float heightOffset;
+        private Vector3 m_StartingPos;
+        
+        void OnEnable()
+        {
+            m_StartingPos = transform.position;
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs:16]: Only the `y` component of m_StartingPos is used, You do not need to store a Vector3, a float should do.

File: Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        }
+
+        void Update()
+        {
+            transform.position = new Vector3(transformToFollow.position.x, m_StartingPos.y + heightOffset, transformToFollow.position.z);
+            Vector3 m_WalkDir = targetToLookAt.position - transform.position;
+            m_WalkDir.y = 0; //flatten dir on the y
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs:23]: This code will only work in flat environments where the `y` axis is not used. (Which is all of our current example environments)
Do you think it is work it to generalize this code?

File: Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+        
+        void OnEnable()
+        {
+            m_StartingPos = transform.position;
+        }
+
+        void Update()
+        {
+            transform.position = new Vector3(transformToFollow.position.x, m_StartingPos.y + heightOffset, transformToFollow.position.z);
+            Vector3 m_WalkDir = targetToLookAt.position - transform.position;
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs:22]: Since this is not a member variable, you do not need `m_` nor capitalization.

File: Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/JointDriveController.cs [hunk 4]
```diff
@@ -122,7 +122,7 @@ public void SetupBodyPart(Transform t)
                 startingPos = t.position,
                 startingRot = t.rotation
             };
-            bp.rb.maxAngularVelocity = 100;
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/JointDriveController.cs:125]: Can we make this value a constant?

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 5]
```diff
@@ -1,16 +1,36 @@
...
 using Unity.MLAgents;
 using Unity.MLAgentsExamples;
 using Unity.MLAgents.Sensors;
+using BodyPart = Unity.MLAgentsExamples.BodyPart;
 
 public class WalkerAgent : Agent
 {
+    [Header("Walking Speed")]
+    [Space(10)]
     [Header("Specific to Walker")]
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:12]: You no longer need this line I think. "Specific to Walker" was to separate the base Agent's fields from the implemented Agent's fields.

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 6]
```diff
@@ -1,16 +1,36 @@
+using MLAgentsExamples;
 using UnityEngine;
 using Unity.MLAgents;
 using Unity.MLAgentsExamples;
 using Unity.MLAgents.Sensors;
+using BodyPart = Unity.MLAgentsExamples.BodyPart;
 
 public class WalkerAgent : Agent
 {
+    [Header("Walking Speed")]
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:10]: The only public variable after this header is `maximumWalkingSpeed`. I think this header will be confusing.

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 7]
```diff
@@ -1,16 +1,36 @@
...
     [Header("Specific to Walker")]
+    public float maximumWalkingSpeed = 999; //The max walk velocity magnitude an agent will be rewarded for
+    Vector3 m_WalkDir; //Direction to the target
+    Quaternion m_WalkDirLookRot; //Will hold the rotation to our target
+    
+    [Space(10)]
+    [Header("Orientation Cube")]
+    [Space(10)]
+    //This will be used as a stable observation platform for the ragdoll to use.
+    GameObject m_OrientationCube;
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:21]: Since this is not public, I don't think it will show in the Inspector, making the Header above confusing.

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 8]
```diff
@@ -1,16 +1,36 @@
...
+    
     [Header("Target To Walk Towards")]
     [Space(10)]
+    public float targetSpawnRadius;
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:25]: Add a comment or a Tooltip

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 9]
```diff
@@ -1,16 +1,36 @@
...
     public Transform target;
+    public Transform ground;
+    public bool detectTargets;
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:28]: These public fields deserve either a more descriptive name or a tooltip/comment to explain what they do

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 10]
```diff
@@ -1,16 +1,36 @@
...
+    public bool targetIsStatic;
+    public bool respawnTargetWhenTouched;
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:30]: I think it should be the target that knows what behavior to have when touched, not the Agent.

From the videos you showed, some Agents can have the same target, if they have different settings, the target's behavior will be ambiguous.

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 11]
```diff
@@ -37,6 +57,14 @@ public class WalkerAgent : Agent
 
     public override void Initialize()
     {
+        //Spawn an orientation cube
+        Vector3 oCubePos = hips.position;
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:61]: Typo ? Why    __o__ CubePos?

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 12]
```diff
@@ -37,6 +57,14 @@ public class WalkerAgent : Agent
...
+        oCubePos.y = -.45f;
+        m_OrientationCube = Instantiate(Resources.Load<GameObject>("OrientationCube"), oCubePos, Quaternion.identity);
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:63]: Isn't it simpler to always have the orientation cube attached since it will always be attached in Initialize anyway ?
This code could break if the "OrientationCube" is renamed (which is not ideal)

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 13]
```diff
@@ -142,37 +168,89 @@ public override void OnActionReceived(float[] vectorAction)
...
         // c. Encourage head height.
-        // d. Discourage head movement.
-        m_DirToTarget = target.position - m_JdController.bodyPartsDict[hips].rb.position;
         AddReward(
-            +0.03f * Vector3.Dot(m_DirToTarget.normalized, m_JdController.bodyPartsDict[hips].rb.velocity)
-            + 0.01f * Vector3.Dot(m_DirToTarget.normalized, hips.forward)
-            + 0.02f * (head.position.y - hips.position.y)
-            - 0.01f * Vector3.Distance(m_JdController.bodyPartsDict[head].rb.velocity,
-                m_JdController.bodyPartsDict[hips].rb.velocity)
+            +0.02f * Vector3.Dot(m_OrientationCube.transform.forward,
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:203]: To clarify this, I suggest naming each component of this reward with a variable.

```csharp
var hipVelocityReward = Vector3.Dot(m_OrientationCube.transform.forward, Vector3.ClampMagnitude(m_JdController.bodyPartsDict[hips].rb.velocity, maximumWalkingSpeed));
var headOrientationReward = Vector3.Dot(m_OrientationCube.transform.forward, head.forward);
var headHeightOverFeetReward = (head.position.y - footL.position.y) + (head.position.y - footR.position.y);

AddReward(
   0.02f * hipVelocityReward + 
   0.01f * headOrientationReward + 
   0.005f * headHeightOverFeetReward
   );
```

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 14]
```diff
@@ -142,37 +168,89 @@ public override void OnActionReceived(float[] vectorAction)
...
         {
-            transform.rotation = Quaternion.LookRotation(m_DirToTarget);
+            GetRandomTargetPos();
         }
+    }
+
+    /// <summary>
+    /// Moves target to a random position within specified radius.
+    /// </summary>
+    public void GetRandomTargetPos()
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:226]: This is not a `Get` Rename to `MoveTargetToRandomPosition()`
I would advise against the Agent moving the target, I think the target should be told it was touched and then  handle its own motion.

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 15]
```diff
@@ -142,37 +168,89 @@ public override void OnActionReceived(float[] vectorAction)
...
         {
             bodyPart.Reset(bodyPart);
         }
+        
+        //Random start rotation to help generalize
+        transform.rotation = Quaternion.Euler(0, Random.Range(0.0f, 360.0f), 0);
+        
+        UpdateOrientationCube();
+
+        if (detectTargets && !targetIsStatic)
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:249]: Not too sure what `detectTargets && !targetIsStatic` does. Can you add a comment for reference.

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 16]
```diff
@@ -187,4 +265,15 @@ public void SetResetParameters()
     {
         SetTorsoMass();
     }
+
+    private void OnDrawGizmosSelected()
+    {
+        if (Application.isPlaying)
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:271]: I __think__ this can only be used in the Editor.

File: Project/ProjectSettings/NetworkManager.asset [hunk 17]
```diff
@@ -0,0 +1,8 @@
+%YAML 1.1
```
Comment (vincentpierre) [inline Project/ProjectSettings/NetworkManager.asset:1]: Why do we need this `.asset` now ?

File: Project/ProjectSettings/ProjectVersion.txt [hunk 18]
```diff
@@ -1 +1 @@
-m_EditorVersion: 2018.4.17f1
```
Comment (vincentpierre) [inline Project/ProjectSettings/ProjectVersion.txt:1]: Are we okay bumping?

File: Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs (hunk 2 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs:23]: yeah i've been thinking about this as well. when we do non-uniform terrain there's more things to consider than flattening the direction here so I was going to wait till we crossed that bridge

File: Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs (hunk 3 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs:22]: ok will fix

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 5 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:12]: will fix

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 14 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:226]: agreed. i think there should be a universal target controller script. we'd need it to work in multiple situations so I think we need a design discussion around this

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 15 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:249]: this is logic related to respawning the target. another artifact from not having a dedicated target controller script.

File: Project/ProjectSettings/NetworkManager.asset (hunk 17 above)
Comment (Hunter-Unity) [reply Project/ProjectSettings/NetworkManager.asset:1]: maybe an artifact. not intentionally added. will remove

File: Project/ProjectSettings/ProjectVersion.txt (hunk 18 above)
Comment (Hunter-Unity) [reply Project/ProjectSettings/ProjectVersion.txt:1]: i thought this was the version we were on. doesn't matter to me which version is used

File: Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs (hunk 1 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs:16]: changed it to float.

File: Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs (hunk 3 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs]: fixed

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 14 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: fixed

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 15 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:248]: added code comment

File: Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Unity.MLAgentsExamples
+{
+    public class DirectionIndicator : MonoBehaviour
+    {
+        public Transform transformToFollow; //ex: hips or body
+        public Transform targetToLookAt; //target in the scene the indicator will point to
+        public float heightOffset;
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs:11]: public fields need a capital case I believe

File: Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/JointDriveController.cs (hunk 4 above)
Comment (vincentpierre) [reply Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/JointDriveController.cs:125]: Something like : 
```csharp
const float k_MaxAngularVelocity = 50.0;
...
bp.rb.maxAngularVelocity = k_MaxAngularVelocity;
```

Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/JointDriveController.cs:125]: done. I changed it to your example. We have other JointDriveController changes to make but i think those should be in a different PR

File: Project/ProjectSettings/NetworkManager.asset (hunk 17 above)
Comment (Hunter-Unity) [reply Project/ProjectSettings/NetworkManager.asset]: removed

File: Project/ProjectSettings/ProjectVersion.txt (hunk 18 above)
Comment (Hunter-Unity) [reply Project/ProjectSettings/ProjectVersion.txt:1]: current master is m_EditorVersion: 2018.4.17f1 so we're good here. no change

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 16 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:269]: I think i may move this logic to a separate script in a future PR

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 7 above)
Comment (vincentpierre) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:20]: @Hunter-Unity bump

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 20]
```diff
@@ -69,36 +95,35 @@ public override void Initialize()
...
         {
-            sensor.AddObservation(bp.currentXNormalizedRot);
-            sensor.AddObservation(bp.currentYNormalizedRot);
-            sensor.AddObservation(bp.currentZNormalizedRot);
+            sensor.AddObservation(bp.rb.transform.localRotation);
             sensor.AddObservation(bp.currentStrength / m_JdController.maxJointForceLimit);
         }
     }
-
+
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:115]: I think formatting should have removed this tab. Why is this here?

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 16 above)
Comment (vincentpierre) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:269]: Does in run in a built executable ?

File: Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs (hunk 19 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/SharedAssets/Scripts/DirectionIndicator.cs:11]: no change. current format is what we're using in other scripts so i'll leave as-is.

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 20 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: removed tab

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 7 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:17]: agreed. i removed 16-18.

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 8 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: added more comments for that section

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 11 above)
Comment (vincentpierre) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:54]: Is it for orientationCubePos ? 
I think it should be spelled out since it is spelled out everywhere else.

File: Project/Assets/ExpertWalkerDyn.demo.meta [hunk 21]
```diff
@@ -1,8 +1,8 @@
 fileFormatVersion: 2
```
Comment (vincentpierre) [inline Project/Assets/ExpertWalkerDyn.demo.meta:1]: I don't think you should have a demo here

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 11 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:54]: removed. refactored to include the cube as part of the prefab

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 12 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:56]: removed. refactored to include the cube as part of the prefab

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 16 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:264]: The OnDrawGizmo callbacks are not executed in builds, only the editor. The builds have worked without any errors.

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 13 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: agreed that looks cleaner. refactored based on your code.

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs (hunk 6 above)
Comment (Hunter-Unity) [reply Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs]: removed the header. when we have more settings to control the walking behavior we can revisit using a header here.

File: Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs [hunk 22]
```diff
@@ -1,17 +1,30 @@
...
+    public float maximumWalkingSpeed = 999; //The max walk velocity magnitude an agent will be rewarded for
+    Vector3 m_WalkDir; //Direction to the target
+    Quaternion m_WalkDirLookRot; //Will hold the rotation to our target
 
-    Vector3 m_DirToTarget;
-    public Transform hips;
+    [Header("Orientation")] [Space(10)]
+    //This will be used as a stable reference point for observations
+    //Because ragdolls can move erratically, using a standalone reference point can significantly improve learning
+    public GameObject orientationCube;
```
Comment (vincentpierre) [inline Project/Assets/ML-Agents/Examples/Walker/Scripts/WalkerAgent.cs:17]: Move this after the body part section.

**Comment by vincentpierre** [comment]
Don't forget to remove the extra demo file !
